design/dedicated-osd-pod.md

```diff
@@ -0,0 +1,66 @@
+
+As observed, all Ceph OSDs are running in the same Pod.
+
+## Limitations
+
+The limitations of current design are:
+
+- Reliability issue. One Pod for all OSDs doesn't have the highest reliability nor efficiency. If the Pod is deleted, accidentally or during maintenence, all OSDs are down till the ReplicaSet restart. 
+- Efficiency issue. Resource limits cannot be set effectively on the OSDs since the number of osds per in the pod could vary from node to node. The operator cannot make decisions about the topology because it doesn't know in advance what devices are available on the nodes.
+- Process management issue. Rook's process management is very simple. Using Kubernetes pod management is much more reliable.
+
+
+A more comprehensive discussions can be found at [this issue](https://github.com/rook/rook/issues/1341).
+
+## Proposal   
+
+We propose the following change to address the limitations.
+
+### Terms
+
+- OSD Discovery. A process that is given a device or directory path filter and use the filter to discover devices on the host. The process updates the OSD orchestration Configmap with the device paths and mark the orchestration status as "DeviceDiscovered". The process is created by Rook Operator and runs in a Kubernetes deployment on the storage node. This process is storage backend agnostic. Currently `rook osd` command implements this functionality.
+
+- OSD Provisioner. A process that is given device or directory paths upon start and make backend specific storage types. For now, the process prepare Ceph OSDs. Currently `rook osd` command implements this functionality.
+
+### Create new OSDs
+| Sequence |Rook Operator | OSD Discovery  | OSD Provisioner   | Ceph OSD Deployment | 
+|---|---|---|---|---|
+| 1  | Read devices from cluster CRD and create an OSD Discovery deployment on each storage node  |   |   |
+| 2  | Watch Configmap  | Discover OSD devices that fit those defined in Cluster CRD and update Configmap with the device paths| | |
+| 3  | Detect Configmap change, parse Configmap, extract device paths, and create an OSD Provisioner deployment for each device  |Detect device additions or removals | | |
+| 4  | Watch Configmap |Detect device additions or removals | Prepare OSDs, Persist OSD ID, datapath, and node info in a per node Configmap | |
+| 5  | Detect Configmap change, parse Configmap, extract OSD info, construct OSD Pod command and arg | Detect device additions or removals | | |
+| 6  | Create one deployment per OSD |Detect device additions or removals | |
+| 7  | |Detect device additions or removals | | Start `ceph-osd` Daemon one Pod per device |
+
+This change addresses the above limitations in the follow ways:
+- High reliability. Each `ceph-osd` daemon runs its own Pod, their restart and upgrade are by Kubernetes controllers. Upgrading OSD Provisioner Pod no longer restarts `ceph-osd` daemons.
```

Comment:
currently, when OSDs are being added/removed, the "provisioning" phase starts OSD daemons so they can participate in rebalancing.  See https://github.com/rook/rook/blob/master/pkg/daemon/ceph/osd/daemon.go#L104

how does this design affect the adding/removing of OSDs over time, waiting for cluster to rebalance, etc.? (especially removal since that's the more difficult case)

Reply:
The operator should have the same level of flexibility to add/remove osds and check for health at each step. Now we will just have greater granularity over the devices and osds instead of at a node level